### PR TITLE
Improve DCE in LIR liveness.

### PR DESCRIPTION
**Pull request by pgavlin**
In particular:
- Rather than only removing dead code as part of dead store removal,
  remove all side-effect-free nodes that either do not produce a value
  or produce a value that is unused.
- When optimizing, set `fgStmtRemoved` in order to indicate that tracked
  lclVar uses or defs have been removed and liveness may need to be
  recalculated. Previously this flag was only set upon eliminating a
  dead store.

### CODE REVIEW

**Comment by pgavlin** [comment]
@dotnet-bot test Tizen armel Cross Debug Build

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT x64 Throughput

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT x64 Throughput

**Comment by pgavlin** [comment]
@dotnet-bot 
test Windows_NT x64 Debug Build and Test
test Windows_NT x64 Throughput

**Comment by pgavlin** [comment]
@BruceForstall @CarolEidt @dotnet/jit-contrib PTAL

Essentially, this performs dead-store removal on SDSU temps as well as LclVars. Throughput results are good: when compared against the [baseline](https://ci2.dot.net/job/dotnet_coreclr/job/perf/job/master/job/perf_throughput_perflab_Windows_NT_x64_full_opt/64/), these changes improve throughput by .1% when using mean times for comparison and regress throughput by the same when using median times for comparison. This lines up pretty well with a reported 0.1% throughput improvement reported by callgrind on `crossgen S.P.C.dll`.

There are a couple small changes to the IR that are necessary in order to enable this:
- Nodes that read the flags register must be marked with `GTF_USE_FLAGS`
- `GTF_IND_REFARR_LAYOUT` has been removed and `GTF_IND_NONFAULTING` updated to use its bit. The former was never used in any case and the latter was overlapping with `GTF_SET_FLAGS`
These changes allow `GTF_SET_FLAGS` to universally indicate that a node writes the flags register. The DCE implementation in this change assumes that any write to the flags register may be observed and will not remove nodes marked with `GTF_SET_FLAGS` (particularly necessary for x86).

I haven't been able to get `jit-diff` running well on my MacBook, so full diffs will have to wait until Monday.

**Comment by pgavlin** [comment]
@dotnet-bot test
OSX10.12 x64 Checked Build and Test
Ubuntu x64 Checked Build and Test
Windows_NT x64 Debug Build and Test

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT x64 Debug Build and Test

**Comment by pgavlin** [comment]
Internal diffs are good. The most common snippets we end up removing are `nop`s and unused compares. The latter is particularly impactful, generally deleting ~19 bytes or so of code:
```
       mov      rcx, rsi
       lea      rdx, [(reloc)]
       cmp      qword ptr [rcx], rdx
       sete     cl
       movzx    rcx, cl
```

Rarer are instances of dead helper calls, e.g. type tests:
```
       xor      rdx, rdx
       mov      rcx, 0xD1FFAB1E
       call     CORINFO_HELP_ISINSTANCEOFCLASS
```

and allocations:
```
       mov      rcx, 0xD1FFAB1E
       call     CORINFO_HELP_NEWSFAST
```

Overall change as reported by SPMI:
```
Base: 8909259
Diff: 8846092
Delta: 63167
Pct. Delta: 0.71
```

**Comment by AndyAyersMS** [comment]
Can you note where you saw the dead isinst/newobj calls? These may be cases we could be cleaning up earlier.

**Comment by mikedn** [comment]
> The latter is particularly impactful, generally deleting ~19 bytes or so of code:

Presumably in the case of the unused compare we keep the indir and remove everything else?

**Comment by pgavlin** [comment]
> Presumably in the case of the unused compare we keep the indir and remove everything else?

We only keep the indir if it might fault. In a number of cases we are able to remove the entire tree, including the indir.

> Can you note where you saw the dead isinst/newobj calls? These may be cases we could be cleaning up earlier.

@russellhadley thought the same. I will take a look.

**Comment by pgavlin** [comment]
@mikedn here's an example where we remove a compare but keep the load around (from `SerializationInfo:set_AssemblyName` in S.P.CoreLib.dll):
```
       cmp      byte  ptr [rsi+78], 0
       sete     cl
       movzx    rcx, cl
```

becomes

```
       movzx    rcx, byte  ptr [rsi+78]
```

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT arm64 Cross Debug Build

File: src/jit/gentree.h [hunk 1]
```diff
@@ -945,9 +946,8 @@ struct GenTree
 #define GTF_INX_REFARR_LAYOUT       0x20000000 // GT_INDEX -- same as GTF_IND_REFARR_LAYOUT
 #define GTF_INX_STRING_LAYOUT       0x40000000 // GT_INDEX -- this uses the special string array layout
 
-#define GTF_IND_NONFAULTING      GTF_SET_FLAGS // GT_IND   -- An indir that cannot fault. GTF_SET_FLAGS is not used on indirs.
+#define GTF_IND_NONFAULTING         0x20000000 // GT_IND   -- An indir that cannot fault.
 #define GTF_IND_VOLATILE            0x40000000 // GT_IND   -- the load or store must use volatile sematics (this is a nop on X86)
```
Comment (BruceForstall) [inline src/jit/gentree.h]: nit: the bits in the per-class sets are generally listed from high value to low

File: src/jit/gentree.cpp [hunk 2]
```diff
@@ -885,6 +885,42 @@ bool GenTreeCall::IsPure(Compiler* compiler) const
            compiler->s_helperCallProperties.IsPure(compiler->eeGetHelperNum(gtCallMethHnd));
 }
 
+bool GenTreeCall::IsSideEffectFree(Compiler* compiler, bool ignoreExceptions, bool ignoreCctors) const
+{
```
Comment (BruceForstall) [inline src/jit/gentree.cpp]: 1. Please add normal function header comment
2. Mention and call out relationship to gtNodeHasSideEffects() -- can that one use this one?
3. Should this instead be named HasSideEffects() to avoid the pseudo-double-negative?

File: src/jit/gentree.h (hunk 1 above)
Comment (BruceForstall) [inline src/jit/gentree.h]: It's disconcerting that (at least) `GTF_IND_VOLATILE/GTF_IND_UNALIGNED` also apply to GT_INDEX, but are not commented as such.

File: src/jit/gentree.cpp (hunk 2 above)
Comment (BruceForstall) [inline src/jit/gentree.cpp]: Re: #2, never mind: I see you changed that func.

---
In reply to: [133565093](https://github.com/dotnet/coreclr/pull/12892#discussion_r133565093) [](ancestors = 133565093)

File: src/jit/liveness.cpp [hunk 3]
```diff
@@ -1687,204 +1687,208 @@ void Compiler::fgComputeLifeCall(VARSET_TP& life, GenTreeCall* call)
...
-//                    legacy backend.
-//    node          - The actual tree node being processed.
-//
-// Returns:
-//    `true` if the local var node corresponds to a dead store; `false`
-//    otherwise.
-//
-bool Compiler::fgComputeLifeLocal(VARSET_TP& life, VARSET_VALARG_TP keepAliveVars, GenTree* lclVarNode, GenTree* node)
+void Compiler::fgComputeLifeTrackedLocalUse(VARSET_TP& life, LclVarDsc& varDsc, GenTreeLclVarCommon* node)
 {
```
Comment (BruceForstall) [inline src/jit/liveness.cpp]: As usual, please add function headers here (and below)

File: src/jit/liveness.cpp [hunk 4]
```diff
@@ -1687,204 +1687,208 @@ void Compiler::fgComputeLifeCall(VARSET_TP& life, GenTreeCall* call)
...
-            // Record interference with other live variables
-            fgMarkIntf(life, varBit);
+    // Record interference with other live variables
+    fgMarkIntf(life, varBit);
+}
+
+//------------------------------------------------------------------------
+// Compiler::fgComputeLifeLocal: compute the changes to local var liveness
+//                               due to a use or a def of a local var and
+//                               indicates wither the use/def is a dead
```
Comment (BruceForstall) [inline src/jit/liveness.cpp]: wither => whether

File: src/jit/liveness.cpp [hunk 5]
```diff
@@ -1687,204 +1687,208 @@ void Compiler::fgComputeLifeCall(VARSET_TP& life, GenTreeCall* call)
...
+//                               store.
+//
+// Arguments:
+//    life          - The live set that is being computed.
+//    keepAliveVars - The currents set of variables to keep alive
```
Comment (BruceForstall) [inline src/jit/liveness.cpp]: currents => current

File: src/jit/liveness.cpp [hunk 6]
```diff
@@ -1687,204 +1687,208 @@ void Compiler::fgComputeLifeCall(VARSET_TP& life, GenTreeCall* call)
...
+    const unsigned varIndex = varDsc.lvVarIndex;
+    if (VarSetOps::IsMember(this, life, varIndex))
+    {
+        /* The variable is live */
+
+        if ((node->gtFlags & GTF_VAR_USEASG) == 0)
         {
-            // Is the variable already known to be alive?
-            if (VarSetOps::IsMember(this, life, varIndex))
+            /* Mark variable as dead from here to its closest use */
```
Comment (CarolEidt) [inline src/jit/liveness.cpp]: Please make this a C++ style comment, while you're here.
Also, I would rephrase this as simply "Remove variable from the live set." Talking about its closest use almost makes one think we're talking about its downstream use.

File: src/jit/liveness.cpp [hunk 7]
```diff
@@ -1687,204 +1687,208 @@ void Compiler::fgComputeLifeCall(VARSET_TP& life, GenTreeCall* call)
...
+    assert(node != nullptr);
 
-            for (unsigned i = varDsc->lvFieldLclStart; i < varDsc->lvFieldLclStart + varDsc->lvFieldCnt; ++i)
-            {
+    if (!varTypeIsStruct(varDsc.lvType) || (lvaGetPromotionType(&varDsc) == PROMOTION_TYPE_NONE))
+    {
+        return;
+    }
+
+    VARSET_TP varBit(VarSetOps::MakeEmpty(this));
```
Comment (CarolEidt) [inline src/jit/liveness.cpp]: Again, I know this is pre-existing code, but `varBit` seems to be a strange name for a set that is likely to have more than one member. I would probably rename this `fieldVars` or something, but IMO it should at least be plural.

File: src/jit/liveness.cpp [hunk 8]
```diff
@@ -1963,17 +1967,207 @@ void Compiler::fgComputeLifeLIR(VARSET_TP& life, BasicBlock* block, VARSET_VALAR
...
+                        }
+
+                        return GenTree::VisitResult::Continue;
+                    });
+
+                    blockRange.Remove(node);
+
+                    // Removing a call does not affect liveness unless it is a tail call in a nethod with P/Invokes or
+                    // is
+                    // itself a P/Invoke, in which case it may affect the liveness of the frame root variable.
```
Comment (CarolEidt) [inline src/jit/liveness.cpp]: Weird comment formatting

File: src/jit/gentree.h (hunk 1 above)
Comment (pgavlin) [inline src/jit/gentree.h]: How is that possible? In the case of `GTF_IND_VOLATILE`, there would be an ambiguity with `GTF_INX_STRING_LAYOUT` were that the case. As far as I can tell these flags only apply to `GT_IND`.

File: src/jit/gentree.cpp [hunk 9]
```diff
@@ -9511,7 +9560,7 @@ void Compiler::gtDispNode(GenTreePtr tree, IndentStack* indentStack, __in __in_z
 
             case GT_IND:
                 // We prefer printing R, V or U
-                if ((tree->gtFlags & (GTF_IND_REFARR_LAYOUT | GTF_IND_VOLATILE | GTF_IND_UNALIGNED)) == 0)
+                if ((tree->gtFlags & (GTF_IND_VOLATILE | GTF_IND_UNALIGNED)) == 0)
```
Comment (briansull) [inline src/jit/gentree.cpp]: The comment above should be updated:

`// We prefer printing V or U`

File: src/jit/gentree.cpp [hunk 10]
```diff
@@ -14079,62 +14128,31 @@ bool Compiler::gtNodeHasSideEffects(GenTreePtr tree, unsigned flags)
     {
         if (tree->OperGet() == GT_CALL)
         {
-            // Generally all GT_CALL nodes are considered to have side-effects.
-            // But we may have a helper call that doesn't have any important side effects.
-            //
-            if (tree->gtCall.gtCallType == CT_HELPER)
+            GenTreeCall* const call             = tree->AsCall();
+            const bool         ignoreExceptions = (flags & GTF_EXCEPT) == 0;
+            const bool         ignoreCctors     = (flags & GTF_IS_IN_CSE) != 0; // We can CSE helpers that run cctors.
```
Comment (briansull) [inline src/jit/gentree.cpp]: I think that we also might be able to CSE expressions that throw exceptions.

Comment (pgavlin) [inline src/jit/gentree.cpp]: That is the case, yes. CSE does not pass the `GTF_EXCEPT` flag to this function.

Comment (briansull) [inline src/jit/gentree.cpp]: We could use a function header for this method describing that the caller passes in `flags` that control the behavior of this method.   It seems a bit odd to me to use the  `flags`  argument on this way.

Alternatively we could pass in two values : 1) ignoreExceptions  and 2 ignoreCctors

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT x86 Checked Build and Test

**Comment by pgavlin** [comment]
@mmitche any idea what's up with the RHEL packages not being found?

**Comment by pgavlin** [comment]
@dotnet-bot test Windows_NT x86 Checked Build and Test

**Comment by pgavlin** [comment]
@dotnet-bot 
test OSX10.12 x64 Checked Build and Test 
test Ubuntu x64 Checked Build and Test 
test Windows_NT arm Cross Checked Build and Test
test Windows_NT arm64 Cross Debug Build
test Windows_NT x64 CoreCLR Perf Tests Correctness
test Windows_NT x64 Debug Build and Test
test Windows_NT x64 Release Priority 1 Build and Test 
test Windows_NT x86 CoreCLR Perf Tests Correctness
